bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -3271,16 +3271,104 @@ private static boolean checkFillerValue(BUnionType type,  List<Type> unAnalyzedT
         if (type.isNullable()) {
             return true;
         }
-        // All members are of same type.
-        Iterator<Type> iterator = type.getMemberTypes().iterator();
-        Type firstMember;
-        for (firstMember = iterator.next(); iterator.hasNext(); ) {
-            if (!isSameType(firstMember, iterator.next())) {
+        return isSameBasicTypeWithFillerValue(type.getMemberTypes());
+    }
+
+    private static boolean isSameBasicTypeWithFillerValue(List<Type> memberTypes) {
+
+        // here finite types and non finite types are separated
+        // for finite types only all their value space items are collected
+        List<Type> nonFiniteTypes = new ArrayList<>();
+        Set<Object> allFiniteTypesValueSpace = new HashSet<>();
```

Comment:
Shall we rename the `allFiniteTypesValueSpace` something like `combinedValueSpace` or `valueSpaceOfFiniteTypes`?